public/app/plugins/datasource/elasticsearch/types.ts

```diff
@@ -32,6 +32,7 @@ interface MetricConfiguration<T extends MetricAggregationType> {
   hasSettings: boolean;
   hasMeta: boolean;
   defaults: Omit<Extract<MetricAggregation, { type: T }>, 'id' | 'type'>;
+  transform?: Record<string, (value: any) => any>;
```

Comment:
@Elfo404 Is there a way to set the types up so the signature of `transform` matches the keys from the `settings` for an aggregation? OR would  we need to create 2 types of `MetricAggregationConfig`?

Reply:
as far as I understood you need to transform what is saved in the query model (in this case `lag`) into a number because Elasticsearch expects it to be a number in the query, right?

while I think the solution is pretty neat (and I'm sure we can manage to make it type safe), this only works when the queries are issued by the frontend and won't support template variables. As we are planning to deprecate browser access mode (and therefore remove the query builder in a future release https://github.com/grafana/grafana/pull/29649) leveraging the plugin backend to build the query.

Given this, I had the idea of doing something similar by using a `format` function (like your `transform` one) in https://github.com/grafana/grafana/blob/82b21fe35e9bbab99bf3c185e17ab6834ef70776/public/app/plugins/datasource/elasticsearch/datasource.ts#L323  for all the `settings`, `meta` etc.. for each bucket and metric aggregation.

The formatting logic would have to be duplicated in the backend though as we would have to save the original string in the query model, which is less than ideal.

My suggestion is to drop that formatting logic for now, and simply treat `lag` as a number (like we already do for the MovingAverage Settings) while we find a better way to handle this in a more generic way, supporting template variables in the frontend and having a generic formatter in the backend, so instead of using `SettingField` something like: 

```tsx
<InlineField label="Lag">
  <Input
    onBlur={e => dispatch(changeMetricSetting(metric, 'lag', parseInt(e.target.value!, 10)))}
    defaultValue={metric.settings?.lag}
  />
</InlineField>
```

Reply:
Sounds good... I went ahead and implemented your suggestion and removed the `transform` stuff.